clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeleteExecutor.scala

```diff
@@ -1,136 +1,159 @@
+      webClient
+        .getMetadataForKey(deleteCommand.index)(deleteCommand.key, includeDeleted = false)
+        .orElse {
+          // Has to be lazy because `orElse` eagerly fails as soon as either branch fails.
+          Source.lazily { () =>
+            Source.failed(
+              new IllegalStateException(
                 s"No $name found in Clio for $prettyKey, nothing to delete."
               )
             )
           }
-        _ <- Future(checkPathConsistency(ioUtil, existingMetadata))
-        upsertResponse <- deleteFiles(webClient, ioUtil, existingMetadata)
-      } yield {
-        upsertResponse
-      }
+        }
     }
   }
 
-  private def checkPathConsistency(
-    ioUtil: IoUtil,
-    metadata: deleteCommand.index.MetadataType
-  ): Unit = {
-    metadata.pathsToDelete.foreach { path =>
-      if (!ioUtil.isGoogleObject(path)) {
-        throw new IllegalStateException(
-          s"Inconsistent state detected: non-cloud path '$path' is registered to the cloud $name for $prettyKey"
-        )
-      }
-    }
-  }
+  /**
+    * Build a stream which, when pulled, will get the paths of all files-to-delete from
+    * the given metadata, then for each path check:
+    *
+    * <ol>
+    * <li>That the path points to cloud storage, and</li>
+    * <li>That the file pointed to by the path actually exists.</li>
+    * </ol>
+    *
+    * If all the checks pass, the stream will emit the list of paths-to-delete.
+    * If not all checks pass, but the delete is being forced, the stream will emit the
+    * list of only those paths which point to existing files.
+    */
+  private def buildDelete(
+    metadata: deleteCommand.index.MetadataType,
+    ioUtil: IoUtil
+  ): Source[immutable.Iterable[URI], NotUsed] = {
 
-  private def deleteFiles(
-    client: ClioWebClient,
-    ioUtil: IoUtil,
-    existingMetadata: deleteCommand.index.MetadataType
-  )(implicit ec: ExecutionContext): Future[UpsertId] = {
-    val pathsToDelete = existingMetadata.pathsToDelete.filter { path =>
-      val pathExists = ioUtil.googleObjectExists(path)
-      val err = s"'$path' associated with $prettyKey does not exist in the cloud."
-      if (!pathExists) {
-        if (!deleteCommand.force) {
-          throw new IllegalStateException(
-            err + " Use --force to mark the record as deleted in Clio"
+    val paths = metadata.pathsToDelete.to[immutable.Iterable]
+    Source[URI](paths)
+      .mapAsync(paths.size) { path =>
+        if (!ioUtil.isGoogleObject(path)) {
+          Future.successful(
+            Left(
+              new IllegalStateException(
+                s"Inconsistent state detected, non-cloud path '$path' registered to $prettyKey"
+              )
+            )
           )
+        } else {
+          Future(Either.catchNonFatal(ioUtil.googleObjectExists(path))).map { attempt =>
+            attempt.flatMap { doesExist =>
+              if (doesExist) {
+                Right(Some(path))
```

Comment:
This should be `Right(Option(path))` just for safety.

Reply:
I disagree! Using `Option(thing)` marks that you think `thing` could be `null`, but if the path was really `null` many other things would have exploded before this point, so I don't want to give the false impression.

Trying to avoid having Zamboni's hyper-paranoid defensive programming styles creep everywhere.

Reply:
Now that I'm squinting at this though, this is really an abuse of `Option` as a pseudo-`Iterable`. I'll see if I can find a better way to express it.